src/lnurl.js

```diff
@@ -0,0 +1,116 @@
+import bech32 from 'bech32'
+import wrap from './lib/promise-wrap'
+
+const debug = require('debug')('lightning-charge')
+
+module.exports = (app, payListen, model, auth) => {
+  const {
+    newInvoice, listInvoicesByLnurlPayEndpoint
+  , getLnurlPayEndpoint, listLnurlPayEndpoints
+  , setLnurlPayEndpoint, delLnurlPayEndpoint
+  } = model
+
+  app.get('/endpoints', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      (await listLnurlPayEndpoints())
+        .map(lnurlpay => addBech23Lnurl(req, lnurlpay))
+    )))
+
+  app.post('/endpoint', auth, wrap(async (req, res) =>
+    res.status(201).send(
+      addBech23Lnurl(req, await setLnurlPayEndpoint(null, req.body))
+    )))
+
+  app.put('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      addBech23Lnurl(req, await setLnurlPayEndpoint(req.params.id, req.body))
+    )))
+
+  app.delete('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(await delLnurlPayEndpoint(req.params.id))))
+
+  app.get('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      addBech23Lnurl(req, await getLnurlPayEndpoint(req.params.id))
+    )))
+
+  app.get('/endpoint/:id/invoices', auth, wrap(async (req, res) =>
+    res.send(await listInvoicesByLnurlPayEndpoint(req.params.id))))
+
+  // this is the actual endpoint users will hit
+  app.get('/lnurl/:id', wrap(async (req, res) => {
+    const lnurlpay = await getLnurlPayEndpoint(req.params.id)
+
+    res.status(200).send({
+      tag: 'payRequest'
+    , minSendable: lnurlpay.min
+    , maxSendable: lnurlpay.max
+    , metadata: makeMetadata(lnurlpay)
+    , commentAllowed: lnurlpay.comment
+    , callback: `https://${req.hostname}/lnurl/${lnurlpay.id}/callback`
+    })
+  }))
+
+  app.get('/lnurl/:id/callback', wrap(async (req, res) => {
+    const lnurlpay = await getLnurlPayEndpoint(req.params.id)
+
+    if (req.query.amount > lnurlpay.max)
+      return res.send({status: 'ERROR', reason: 'amount too large'})
+    if (req.query.amount < lnurlpay.min)
+      return res.send({status: 'ERROR', reason: 'amount too small'})
+
+    let invoiceMetadata = {...req.query}
+    delete invoiceMetadata.amount
+    delete invoiceMetadata.fromnodes
+    delete invoiceMetadata.nonce
```

Comment:
The rationale for keeping the other query parameters is the following:

Imagine a shop like https://bitclouds.sh/. Every VM you buy there gets its own lnurlpay you can use to recharge it (not really true today but it's the main use case I have in mind and one that I expect to adopt this charge-lnurlpay feature once their refactor is ready).

So instead of asking Charge to create a new lnurlpay endpoint every time someone creates a VM, Bitclouds can pregenerate a single lnurlpay endpoint and modify its URL by adding an identifier for each VM.

If the base URL is `https://charge.bitclouds.sh/lnurl/recharge` it can then assign `https://charge.bitclouds.sh/lnurl/recharge?vm=nusakan58` to the owner of a VM called `nusakan58`. Then every time a payment is received at that Bitclouds will get a webhook containing `{"vm": "nusakan58"}` and know what to do with it.

Reply:
But your comment about the comment is very ~commentive~ good.